### PR TITLE
Allow the user database to have multiple encryption algorithms

### LINKED ISSUE #48: Add algorithm sha1 in available algorithms
In available algorithms we should still have the algorithm `sha1` for older projects.
We probably want to display a warning if the user set the algorithm to `sha1` and add a warning in documentation.

### PR DESCRIPTION
**Pull request by scottinet**
# Description

* Reintroduce SHA-1 to the list of supported encryption algorithms, allowing to import users and still allow them to log in (fix #48)
* For the same reason, allow passwords to be encrypted without key stretching
* User documents now contain the algorithm name used to encrypt the password, and if a key stretching algorithm has been applied to it or not. If one of these properties are absent from a user document, the current plugin configuration is applied instead
* Now able to check user passwords encrypted with a different algorithm and key stretching configuration than the ones provided in the settings. If the password check succeeds, then it gets re-encrypted using the current plugin configuration, allowing to bring it on par with the current security configuration
* A `pepper` property is now added to every user documents. It is always set to `false`. This property will allow us to add a pepper value to the password later, without introducing any breaking change (see #50)

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -147,25 +165,34 @@ class AuthenticationPlugin {
...
+
     return this.getUsersRepository().get(username)
       .then(result => {
         if (result === null) {
-          return Promise.resolve({
-            kuid: null,
-            message: 'wrong username or password'
-          });
+          // no user found: directly return a password check fail
+          return false;
```
Comment (ballinette) [inline lib/index.js]: nitpicking: should apply the same changes in line 183 below as well

File: lib/index.js [hunk 2]
```diff
@@ -147,25 +165,34 @@ class AuthenticationPlugin {
...
           return Promise.resolve({
             kuid: null,
             message: 'wrong username or password'
           });
         }
 
-        return Promise.resolve({
-          kuid: result.kuid
-        });
+        if (user.algorithm !== this.config.algorithm || user.stretching !== this.config.stretching) {
```
Comment (ballinette) [inline lib/index.js]: :+1: <3

File: lib/passwordManager.js [hunk 3]
```diff
@@ -31,21 +35,42 @@ class PasswordManager {
...
+    const keylen = this.availableAlgorithms[algorithm];
+
+    if (keylen === undefined) {
+      // encryption algorithm deliberately absent from the error message
+      return Promise.reject(new this.context.errors.BadRequestError('Unknown encryption algorithm'));
+    }
+
+    return new Promise((resolve, reject) => {
+      if (stretching) {
+        crypto.pbkdf2(password, salt, 10000, keylen, this.config.algorithm, (err, key) => {
```
Comment (ballinette) [inline lib/passwordManager.js]: sed s/this.config.algorithm/algorithm/

Comment (scottinet) [inline lib/passwordManager.js]: Ouch!
Fixed.

**Pull request merged**